File: archive/cdk-python/Pr6927/lib/api_gateway_stack.py
"""api_gateway_stack.py
API Gateway stack with consolidated REST APIs.
"""

from typing import Optional
import aws_cdk as cdk
from aws_cdk import (
    aws_apigateway as apigw,
    aws_lambda as _lambda,
    aws_logs as logs,
)
from constructs import Construct


class ApiGatewayStackProps:
    """Properties for API Gateway Stack."""

    def __init__(
        self,
        environment_suffix: str,
        environment: str,
        payment_processor: _lambda.IFunction,
        transaction_validator: _lambda.IFunction,
        fraud_detector: _lambda.IFunction
    ):
        self.environment_suffix = environment_suffix
        self.environment = environment
        self.payment_processor = payment_processor
        self.transaction_validator = transaction_validator
        self.fraud_detector = fraud_detector


class ApiGatewayStack(cdk.Stack):
    """
    API Gateway Stack consolidating multiple REST APIs into a single deployment.
    Requirement 3: Consolidate duplicate API Gateway REST APIs into a single deployment
    Requirement 6: Configure CloudWatch Log Groups with 7-day retention
    """

    def __init__(
        self,
        scope: Construct,
        construct_id: str,
        props: ApiGatewayStackProps,
        **kwargs
    ):
        super().__init__(scope, construct_id, **kwargs)

        environment_suffix = props.environment_suffix
        environment = props.environment

        # Cost allocation tags
        tags = {
            "Environment": environment,
            "Team": "payments",
            "CostCenter": "engineering",
            "Project": "payment-processing"
        }

        # Create CloudWatch Log Group with 7-day retention (Requirement 6)
        log_group = logs.LogGroup(
            self,
            f"{environment}-payment-log-api",
            # 7-day retention (Requirement 6)
            retention=logs.RetentionDays.ONE_WEEK,
            removal_policy=cdk.RemovalPolicy.DESTROY
        )

        # Consolidated REST API (Requirement 3)
        # Instead of multiple separate APIs, create one unified API
        self.api = apigw.RestApi(
            self,
            f"{environment}-payment-api-main",
            rest_api_name=f"{environment}-payment-api-consolidated",
            description="Consolidated payment processing API",
            deploy_options=apigw.StageOptions(
                stage_name=environment,
                metrics_enabled=True,
                logging_level=apigw.MethodLoggingLevel.INFO,
                data_trace_enabled=True,
                access_log_destination=apigw.LogGroupLogDestination(log_group),
                access_log_format=apigw.AccessLogFormat.json_with_standard_fields(
                    caller=True,
                    http_method=True,
                    ip=True,
                    protocol=True,
                    request_time=True,
                    resource_path=True,
                    response_length=True,
                    status=True,
                    user=True)))

        # Payments resource and methods
        payments = self.api.root.add_resource("payments")
        payments.add_method(
            "POST",
            apigw.LambdaIntegration(props.payment_processor)
        )

        # Transactions resource and methods
        transactions = self.api.root.add_resource("transactions")
        transactions.add_method(
            "POST",
            apigw.LambdaIntegration(props.transaction_validator)
        )
        transactions.add_method(
            "GET",
            apigw.LambdaIntegration(props.transaction_validator)
        )

        # Fraud detection resource and methods
        fraud = self.api.root.add_resource("fraud")
        fraud.add_method(
            "POST",
            apigw.LambdaIntegration(props.fraud_detector)
        )

        # Apply cost allocation tags
        for key, value in tags.items():
            cdk.Tags.of(self.api).add(key, value)

        # Outputs
        cdk.CfnOutput(
            self,
            "ApiUrl",
            value=self.api.url,
            export_name=f"{environment}-payment-api-url"
        )

        cdk.CfnOutput(
            self,
            "ApiId",
            value=self.api.rest_api_id,
            export_name=f"{environment}-payment-api-id"
        )

File: archive/cdk-ts/Pr4356/lib/stacks/api-gateway-stack.ts
import * as cdk from 'aws-cdk-lib';
import * as apigateway from 'aws-cdk-lib/aws-apigateway';
import * as lambda from 'aws-cdk-lib/aws-lambda';
import * as logs from 'aws-cdk-lib/aws-logs';
import { Construct } from 'constructs';

interface ApiGatewayStackProps extends cdk.StackProps {
  environmentSuffix: string;
  dataProcessorFunction: lambda.Function;
}

export class ApiGatewayStack extends cdk.Stack {
  public readonly api: apigateway.RestApi;

  constructor(scope: Construct, id: string, props: ApiGatewayStackProps) {
    super(scope, id, props);

    // Apply tags to all resources in this stack
    cdk.Tags.of(this).add('iac-rlhf-amazon', 'true');
    cdk.Tags.of(this).add('Environment', props.environmentSuffix);

    // CloudWatch Log Group for API Gateway
    const apiLogGroup = new logs.LogGroup(
      this,
      `ApiLogGroup-${props.environmentSuffix}`,
      {
        logGroupName: `/aws/apigateway/serverless-api-${props.environmentSuffix}`,
        retention: logs.RetentionDays.ONE_MONTH,
        removalPolicy: cdk.RemovalPolicy.DESTROY,
      }
    );

    // REST API
    this.api = new apigateway.RestApi(this, `Api-${props.environmentSuffix}`, {
      restApiName: `serverless-api-${props.environmentSuffix}`,
      description: 'Serverless Data Processing API',
      deployOptions: {
        stageName: props.environmentSuffix,
        tracingEnabled: true,
        dataTraceEnabled: true,
        loggingLevel: apigateway.MethodLoggingLevel.INFO,
        accessLogDestination: new apigateway.LogGroupLogDestination(
          apiLogGroup
        ),
        accessLogFormat: apigateway.AccessLogFormat.jsonWithStandardFields({
          caller: true,
          httpMethod: true,
          ip: true,
          protocol: true,
          requestTime: true,
          resourcePath: true,
          responseLength: true,
          status: true,
          user: true,
        }),
        metricsEnabled: true,
        throttlingBurstLimit: 100,
        throttlingRateLimit: 50,
      },
      defaultCorsPreflightOptions: {
        allowOrigins: apigateway.Cors.ALL_ORIGINS,
        allowMethods: apigateway.Cors.ALL_METHODS,
        allowHeaders: [
          'Content-Type',
          'X-Amz-Date',
          'Authorization',
          'X-Api-Key',
          'X-Amz-Security-Token',
        ],
        allowCredentials: true,
        maxAge: cdk.Duration.days(1),
      },
      endpointTypes: [apigateway.EndpointType.REGIONAL],
    });

    // Request Validator
    const requestValidator = new apigateway.RequestValidator(
      this,
      `RequestValidator-${props.environmentSuffix}`,
      {
        restApi: this.api,
        requestValidatorName: 'validate-body',
        validateRequestBody: true,
        validateRequestParameters: false,
      }
    );

    // Request Model
    const dataModel = new apigateway.Model(
      this,
      `DataModel-${props.environmentSuffix}`,
      {
        restApi: this.api,
        contentType: 'application/json',
        modelName: 'DataModel',
        schema: {
          type: apigateway.JsonSchemaType.OBJECT,
          required: ['data'],
          properties: {
            data: {
              type: apigateway.JsonSchemaType.OBJECT,
            },
            metadata: {
              type: apigateway.JsonSchemaType.OBJECT,
            },
          },
        },
      }
    );

    // Lambda Integration
    const lambdaIntegration = new apigateway.LambdaIntegration(
      props.dataProcessorFunction,
      {
        requestTemplates: { 'application/json': '{ "statusCode": "200" }' },
        integrationResponses: [
          {
            statusCode: '200',
            responseParameters: {
              'method.response.header.Access-Control-Allow-Origin': "'*'",
            },
          },
          {
            statusCode: '500',
            selectionPattern: '.*Error.*',
            responseParameters: {
              'method.response.header.Access-Control-Allow-Origin': "'*'",
            },
          },
        ],
      }
    );

    // API Resources and Methods
    const dataResource = this.api.root.addResource('data');

    // POST /data
    dataResource.addMethod('POST', lambdaIntegration, {
      requestValidator,
      requestModels: {
        'application/json': dataModel,
      },
      methodResponses: [
        {
          statusCode: '200',
          responseParameters: {
            'method.response.header.Access-Control-Allow-Origin': true,
          },
        },
        {
          statusCode: '500',
          responseParameters: {
            'method.response.header.Access-Control-Allow-Origin': true,
          },
        },
      ],
    });

    // GET /data (for health check)
    dataResource.addMethod('GET', lambdaIntegration, {
      methodResponses: [
        {
          statusCode: '200',
          responseParameters: {
            'method.response.header.Access-Control-Allow-Origin': true,
          },
        },
      ],
    });

    // API Key and Usage Plan
    const apiKey = new apigateway.ApiKey(
      this,
      `ApiKey-${props.environmentSuffix}`,
      {
        apiKeyName: `serverless-api-key-${props.environmentSuffix}`,
        description: 'API Key for serverless application',
      }
    );

    const usagePlan = new apigateway.UsagePlan(
      this,
      `UsagePlan-${props.environmentSuffix}`,
      {
        name: `serverless-usage-plan-${props.environmentSuffix}`,
        throttle: {
          rateLimit: 100,
          burstLimit: 200,
        },
        quota: {
          limit: 10000,
          period: apigateway.Period.DAY,
        },
        apiStages: [
          {
            api: this.api,
            stage: this.api.deploymentStage,
          },
        ],
      }
    );

    usagePlan.addApiKey(apiKey);

    // Outputs
    new cdk.CfnOutput(this, 'ApiEndpoint', {
      value: this.api.url,
      description: 'API Gateway endpoint URL',
      exportName: `ApiEndpoint-${props.environmentSuffix}`,
    });

    new cdk.CfnOutput(this, 'ApiKeyId', {
      value: apiKey.keyId,
      description: 'API Key ID',
      exportName: `ApiKeyId-${props.environmentSuffix}`,
    });
  }
}
